nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-documentation/src/main/java/org/apache/nifi/documentation/html/AbstractHtmlDocumentationWriter.java

```diff
@@ -0,0 +1,356 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nifi.documentation.html;
+
+import org.apache.nifi.documentation.DocumentationWriter;
+import org.apache.nifi.util.StringUtils;
+
+import javax.xml.stream.FactoryConfigurationError;
+import javax.xml.stream.XMLOutputFactory;
+import javax.xml.stream.XMLStreamException;
+import javax.xml.stream.XMLStreamWriter;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+abstract class AbstractHtmlDocumentationWriter<T> implements DocumentationWriter<T> {
+
+    /**
+     * The filename where additional user specified information may be stored.
+     */
+    public static final String ADDITIONAL_DETAILS_HTML = "additionalDetails.html";
+
+    static final String NO_DESCRIPTION = "No description provided.";
+    static final String NO_TAGS = "No tags provided.";
+    static final String NO_PROPERTIES = "This component has no required or optional properties.";
+
+    static final String H2 = "h2";
+    static final String H3 = "h3";
+    static final String H4 = "h4";
+    static final String P = "p";
+    static final String BR = "br";
+    static final String SPAN = "span";
+    static final String STRONG = "strong";
+    static final String TABLE = "table";
+    static final String TH = "th";
+    static final String TR = "tr";
+    static final String TD = "td";
+    static final String UL = "ul";
+    static final String LI = "li";
+    static final String ID = "id";
+
+    @Override
+    public void write(final T component, final OutputStream streamToWriteTo, final boolean includesAdditionalDocumentation) throws IOException {
```

Comment:
Minor naming, recommend renaming the stream variable to simply `outputStream`:
```suggestion
    public void write(final T component, final OutputStream outputStream, final boolean includesAdditionalDocumentation) throws IOException {
```